File: q4_trust_layer/proxy/internal/canonical/canonical.go
// Package canonical is a Go port of schema/canonical.py. It MUST stay behaviourally
// identical to that file (see schema/golden_vectors.json for the cross-language
// parity contract) because the proxy and the Python control/harness code compute
// def_hash and event_hash independently and compare them.
//
// Canonical form = RFC 8785 (JSON Canonicalization Scheme, "JCS"). Rather than
// re-implement JCS's UTF-16 key sorting and ECMAScript number formatting by hand
// (the two easiest places to get a byte-for-byte contract subtly wrong), we lean
// on github.com/gowebpki/jcs, a dedicated RFC 8785 implementation: we marshal the
// Go value to JSON with the stdlib encoder, then hand those bytes to jcs.Transform,
// which re-parses and re-serializes per the RFC. That is the documented allowed
// dependency (see q4_trust_layer/DESIGN.md and DECISIONS.md, "Go port of canonical
// hashing").
package canonical

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"sort"
	"unicode/utf16"

	"github.com/gowebpki/jcs"
)

// HashPrefix matches canonical.py's HASH_PREFIX.
const HashPrefix = "sha256:"

// CanonicalJSON serializes v per RFC 8785, matching canonical.py's canonical_json.
func CanonicalJSON(v any) (string, error) {
	raw, err := json.Marshal(v)
	if err != nil {
		return "", fmt.Errorf("canonical: value is not JSON-serializable: %w", err)
	}
	out, err := jcs.Transform(raw)
	if err != nil {
		return "", fmt.Errorf("canonical: JCS transform failed: %w", err)
	}
	return string(out), nil
}

// Sha256Hex matches canonical.py's sha256_hex: HASH_PREFIX + sha256(canonical_json(v)).
func Sha256Hex(v any) (string, error) {
	s, err := CanonicalJSON(v)
	if err != nil {
		return "", err
	}
	sum := sha256.Sum256([]byte(s))
	return HashPrefix + hex.EncodeToString(sum[:]), nil
}

// normalizeSchema ports canonical.py's _normalize_schema: drop $comment at every
// level, and sort+dedupe any "required" array of strings by UTF-16 code unit.
func normalizeSchema(node any) any {
	switch n := node.(type) {
	case map[string]any:
		out := make(map[string]any, len(n))
		for k, v := range n {
			if k == "$comment" {
				continue
			}
			out[k] = normalizeSchema(v)
		}
		if req, ok := out["required"]; ok {
			if arr, ok := req.([]any); ok {
				if strs, ok := allStrings(arr); ok {
					out["required"] = sortDedupeUTF16(strs)
				}
			}
		}
		return out
	case []any:
		out := make([]any, len(n))
		for i, v := range n {
			out[i] = normalizeSchema(v)
		}
		return out
	default:
		return node
	}
}

func allStrings(arr []any) ([]string, bool) {
	out := make([]string, len(arr))
	for i, v := range arr {
		s, ok := v.(string)
		if !ok {
			return nil, false
		}
		out[i] = s
	}
	return out, true
}

// sortDedupeUTF16 sorts unique strings by UTF-16 code unit, matching Python's
// `sorted(set(req), key=lambda s: s.encode("utf-16-be"))`.
func sortDedupeUTF16(strs []string) []any {
	seen := make(map[string]bool, len(strs))
	uniq := make([]string, 0, len(strs))
	for _, s := range strs {
		if !seen[s] {
			seen[s] = true
			uniq = append(uniq, s)
		}
	}
	sort.Slice(uniq, func(i, j int) bool { return utf16Less(uniq[i], uniq[j]) })
	out := make([]any, len(uniq))
	for i, s := range uniq {
		out[i] = s
	}
	return out
}

func utf16Less(a, b string) bool {
	ua := utf16.Encode([]rune(a))
	ub := utf16.Encode([]rune(b))
	n := len(ua)
	if len(ub) < n {
		n = len(ub)
	}
	for i := 0; i < n; i++ {
		if ua[i] != ub[i] {
			return ua[i] < ub[i]
		}
	}
	return len(ua) < len(ub)
}

// NormalizeTool ports canonical.py's normalize_tool: project a tool definition
// onto the fields that define its contract. description is included on purpose
// (a changed description is exactly what a rug-pull looks like); whether the
// change *matters* is the differ's job, not the hot path's.
func NormalizeTool(tool map[string]any) map[string]any {
	out := map[string]any{
		"name":        tool["name"],
		"description": "",
		"inputSchema": normalizeSchema(tool["inputSchema"]),
	}
	if d, ok := tool["description"]; ok {
		out["description"] = d
	}
	for _, k := range [...]string{"outputSchema", "annotations"} {
		if v, ok := tool[k]; ok {
			out[k] = normalizeSchema(v)
		}
	}
	return out
}

// ToolHash matches canonical.py's tool_hash.
func ToolHash(tool map[string]any) (string, error) {
	return Sha256Hex(NormalizeTool(tool))
}

// EventHash matches canonical.py's event_hash: hash of an event for the
// per-session hash chain, excluding the event's own event_hash field.
func EventHash(event map[string]any) (string, error) {
	filtered := make(map[string]any, len(event))
	for k, v := range event {
		if k == "event_hash" {
			continue
		}
		filtered[k] = v
	}
	return Sha256Hex(filtered)
}

File: q3_tool_harness/harness/_canonical.py
"""Reference canonicalization + hashing for tool definitions and recording events.

This file is the cross-language contract. Python (q3 harness, q4 control) and Go
(q4 proxy) must produce byte-identical canonical forms, verified by
`golden_vectors.json`. Stdlib only, so it can be vendored anywhere.

Canonical form = RFC 8785 (JSON Canonicalization Scheme, "JCS"):
  * object keys sorted by UTF-16 code units, no insignificant whitespace
  * strings escaped the way ECMAScript JSON.stringify does
  * numbers serialized the way ECMAScript Number.prototype.toString does

Tool-definition hashing applies a *minimal* normalization first (drop `$comment`,
sort + dedupe `required`). Anything smarter (resolving $ref, deciding a
description edit is cosmetic) belongs in the differ, off the hot path, so the Go
proxy never has to reimplement it.
"""

from __future__ import annotations

import hashlib
import math
from decimal import Decimal
from typing import Any

HASH_PREFIX = "sha256:"

_ESCAPES = {'"': '\\"', "\\": "\\\\", "\b": "\\b", "\f": "\\f", "\n": "\\n", "\r": "\\r", "\t": "\\t"}


def _encode_string(s: str) -> str:
    out = ['"']
    for ch in s:
        if ch in _ESCAPES:
            out.append(_ESCAPES[ch])
        elif ord(ch) < 0x20:
            out.append(f"\\u{ord(ch):04x}")
        elif 0xD800 <= ord(ch) <= 0xDFFF:
            raise ValueError("lone surrogate is not valid in canonical JSON")
        else:
            out.append(ch)
    out.append('"')
    return "".join(out)


def _encode_number(n: int | float) -> str:
    if isinstance(n, bool):  # bool is a subclass of int
        raise TypeError("bool is not a number")
    if isinstance(n, int):
        if abs(n) > 2**53:
            raise ValueError("integer outside IEEE-754 safe range; not portable across languages")
        return str(n)
    if not math.isfinite(n):
        raise ValueError("NaN/Infinity are not valid JSON")
    if n == 0:
        return "0"  # also covers -0.0
    # repr() yields the shortest round-tripping digits, same as ECMAScript.
    sign, digits, exp = Decimal(repr(n)).normalize().as_tuple()
    ds = "".join(map(str, digits))
    k = len(ds)
    e = exp + k  # value = 0.ds * 10^e  (ECMAScript's "n")
    neg = "-" if sign else ""
    if k <= e <= 21:
        body = ds + "0" * (e - k)
    elif 0 < e <= 21:
        body = ds[:e] + "." + ds[e:]
    elif -6 < e <= 0:
        body = "0." + "0" * (-e) + ds
    else:
        mant = ds[0] + ("." + ds[1:] if k > 1 else "")
        x = e - 1
        body = f"{mant}e{'+' if x > 0 else '-'}{abs(x)}"
    return neg + body


def canonical_json(value: Any) -> str:
    """Serialize `value` per RFC 8785."""
    if value is None:
        return "null"
    if value is True:
        return "true"
    if value is False:
        return "false"
    if isinstance(value, (int, float)):
        return _encode_number(value)
    if isinstance(value, str):
        return _encode_string(value)
    if isinstance(value, (list, tuple)):
        return "[" + ",".join(canonical_json(v) for v in value) + "]"
    if isinstance(value, dict):
        for k in value:
            if not isinstance(k, str):
                raise TypeError(f"object keys must be strings, got {type(k).__name__}")
        items = sorted(value.items(), key=lambda kv: kv[0].encode("utf-16-be"))
        return "{" + ",".join(_encode_string(k) + ":" + canonical_json(v) for k, v in items) + "}"
    raise TypeError(f"not JSON-serializable: {type(value).__name__}")


def sha256_hex(value: Any) -> str:
    return HASH_PREFIX + hashlib.sha256(canonical_json(value).encode("utf-8")).hexdigest()


def _normalize_schema(node: Any) -> Any:
    if isinstance(node, dict):
        out = {k: _normalize_schema(v) for k, v in node.items() if k != "$comment"}
        req = out.get("required")
        if isinstance(req, list) and all(isinstance(r, str) for r in req):
            out["required"] = sorted(set(req), key=lambda s: s.encode("utf-16-be"))
        return out
    if isinstance(node, list):
        return [_normalize_schema(v) for v in node]
    return node


def normalize_tool(tool: dict[str, Any]) -> dict[str, Any]:
    """Project a tool definition onto the fields that define its contract.

    Description IS included: a changed description is exactly what a rug-pull
    looks like, so it must change the hash. Whether the change *matters* is the
    differ's job.
    """
    out: dict[str, Any] = {
        "name": tool["name"],
        "description": tool.get("description", ""),
        "inputSchema": _normalize_schema(tool.get("inputSchema", {})),
    }
    for optional in ("outputSchema", "annotations"):
        if optional in tool:
            out[optional] = _normalize_schema(tool[optional])
    return out


def tool_hash(tool: dict[str, Any]) -> str:
    return sha256_hex(normalize_tool(tool))


def event_hash(event: dict[str, Any]) -> str:
    """Hash of an event for the per-session hash chain (excludes its own hash)."""
    return sha256_hex({k: v for k, v in event.items() if k != "event_hash"})
